core/market/decentralized/tests/test_agreement.rs

```diff
@@ -43,3 +44,330 @@ async fn test_gsb_get_agreement() -> Result<(), anyhow::Error> {
+}
+
+#[cfg_attr(not(feature = "market-test-suite"), ignore)]
+#[actix_rt::test]
+async fn approval_before_agreement_confirmation_should_fail() -> anyhow::Result<()> {
+    let req = "Node-1";
+    let prov = "Node-2";
+    let network = MarketsNetwork::new("approval_before_agreement_confirmation_should_fail")
+        .await
+        .add_market_instance(req)
+        .await?
+        .add_market_instance(prov)
+        .await?;
+
+    let proposal_id = exchange_draft_proposals(&network, req, prov).await?;
+    let req_market = network.get_market(req);
+    let req_id = network.get_default_id(req);
+    let prov_id = network.get_default_id(prov);
+
+    // Requestor creates agreement with 1h expiration
+    let agreement_id = req_market
+        .requestor_engine
+        .create_agreement(
+            req_id.clone(),
+            &proposal_id,
+            Utc::now() + Duration::hours(1),
+        )
+        .await?;
+
+    // Provider tries to approve the Agreement
+    let result = network
+        .get_market(prov)
+        .provider_engine
+        .approve_agreement(prov_id.clone(), &agreement_id, 0.1)
+        .await;
+
+    // ... which results in not found error, bc there was no confirmation
+    // so Requestor did not send an Agreement
+    assert_eq!(
+        result.unwrap_err().to_string(),
+        AgreementError::NotFound(agreement_id).to_string()
+    );
+
+    Ok(())
+}
+
+#[cfg_attr(not(feature = "market-test-suite"), ignore)]
+#[actix_rt::test]
+async fn approval_without_waiting_for_agreement_should_pass() -> anyhow::Result<()> {
+    let req = "Node-1";
+    let prov = "Node-2";
+    let network = MarketsNetwork::new("approval_without_waiting_for_agreement_should_pass")
+        .await
+        .add_market_instance(req)
+        .await?
+        .add_market_instance(prov)
+        .await?;
+
+    let proposal_id = exchange_draft_proposals(&network, req, prov).await?;
+    let req_market = network.get_market(req);
+    let req_id = network.get_default_id(req);
+    let prov_id = network.get_default_id(prov);
+
+    // Requestor creates agreement with 1h expiration
+    let agreement_id = req_market
+        .requestor_engine
+        .create_agreement(
+            req_id.clone(),
+            &proposal_id,
+            Utc::now() + Duration::hours(1),
+        )
+        .await?;
+
+    // Confirms it immediately
+    req_market
+        .requestor_engine
+        .confirm_agreement(req_id.clone(), &agreement_id)
+        .await?;
+
+    // Provider successfully approves the Agreement
+    // even though Requestor does not wait for it
+    network
+        .get_market(prov)
+        .provider_engine
+        .approve_agreement(prov_id.clone(), &agreement_id, 0.1)
+        .await?;
+
+    // TODO: is it really ok to allow such situation?
+    // shouldn't we return an error here?
+    // assert_eq!(
+    //     result.unwrap_err().to_string(),
+    //     AgreementError::ProtocolApprove(..).to_string()
+    // );
+
```

Comment:
> Not only allowed, but we must do it this way because otherwise, we will have races. Provider shouldn't depend on Requestor behavior. And if we don't allow to approve, before wait call on requestor side, Provider must sleep until Requestor will call it??? It would be horrible from Provider implementer perspective.

I'm not so sure we must. Both provider and requestor specifies `timeout`. One of timeout reasons is that the other side is not called approval/rejection. But Requestor's side is not symmetrical to Provider's here, since Requestor need to: 
1. create agreement
1. **confirm** it 
1. wait for approval

Confirmation is blocking approval on Provider side. This is enough. I can sleep soundly ;)

> I think you could check in this test `wait_for_approval` after `approve_agreement` scenario, which I by the way implemented (remember this loop in implementation? It was for this case)

ok!

Reply:
But still I don't think Provider should get error, if it called approve earlier.

Reply:
done